src/pudl/transform/ferc1.py

```diff
@@ -124,6 +124,117 @@ class Config:
 ################################################################################
 
 
+################################################################################
+# FERC 1 transform helper functions. Probably to be integrated into a class
+# below as methods or moved to a different module once it's clear where they belong.
+################################################################################
+def get_ferc1_dbf_rows_to_map(ferc1_engine: sa.engine.Engine) -> pd.DataFrame:
+    """Identify DBF rows that need to be mapped to XBRL columns.
+
+    Select all records in the ``f1_row_lit_tbl`` where the row literal associated with a
+    given combination of table and row number is different from the preceeding year.
+    This is the smallest set of records which we can use to reproduce the whole table by
+    expanding the time series to include all years, and forward filling the row
+    literals.
+    """
+    idx_cols = ["sched_table_name", "row_status", "row_number", "report_year"]
+    data_cols = ["row_literal"]
+    row_lit = pd.read_sql(
+        "f1_row_lit_tbl", con=ferc1_engine, columns=idx_cols + data_cols
+    ).sort_values(idx_cols)
+    row_lit["shifted"] = row_lit.groupby(
+        ["sched_table_name", "row_status", "row_number"]
+    ).row_literal.shift()
+    row_lit["changed"] = row_lit.row_literal != row_lit.shifted
+    return row_lit.loc[row_lit.changed, idx_cols + data_cols]
+
+
+def update_ferc1_dbf_xbrl_glue(ferc1_engine: sa.engine.Engine) -> pd.DataFrame:
+    """Regenerate the FERC 1 DBF+XBRL glue while retaining existing mappings.
+
+    Reads all rows that need to be mapped out of the ``f1_row_lit_tbl`` and
+    appends columns containing any previously mapped values, returning the
+    resulting dataframe.
+
+    This
+    """
+    idx_cols = ["sched_table_name", "row_status", "row_number", "report_year"]
+    all_rows = get_ferc1_dbf_rows_to_map(ferc1_engine).set_index(idx_cols)
+    with importlib.resources.open_text(
+        "pudl.package_data.ferc1", "dbf_to_xbrl.csv"
+    ) as file:
+        mapped_rows = (
+            pd.read_csv(file).set_index(idx_cols).drop(["row_literal"], axis="columns")
+        )
+    return (
+        pd.concat([all_rows, mapped_rows], axis="columns")
+        .reset_index()
+        .sort_values(["sched_table_name", "row_status", "report_year", "row_number"])
+    )
+
+
+def get_ferc1_dbf_xbrl_glue(dbf_table_name: str) -> pd.DataFrame:
+    """Read the DBF+XBRL glue and expand it to cover all DBF years."""
+    with importlib.resources.open_text(
+        "pudl.package_data.ferc1", "dbf_to_xbrl.csv"
+    ) as file:
+        row_map = pd.read_csv(
+            file,
+            usecols=[
+                "sched_table_name",
+                "row_status",
+                "report_year",
+                "row_number",
+                "row_type",
+                "xbrl_column_stem",
```

Comment:
The suffix here (for the `plant_in_service` table) indicates what type of value is being reported (starting/ending balance, additions, retirements...) and plays no role in the DBF to XBRL mapping process. Each DBF row number corresponds to all 6 columns in DBF, and we are recreating that structure in the XBRL.

The column names don't let us infer anything about the hierarchy of categories (it looks like you can, almost, sometimes, but nope!). I don't know what we'll end up calling this, since I don't know if other tables will have a similar `stem` + `suffix` structure in the names.

Reply:
ah i get it. and i can see that this changing depending on the new and different things in the other tables. if it were me i'd either change the name of `stem` to be more descriptive or at the very least add this comment somewhere to remind ourselves of wft this means.

Reply:
When I write up instructions for the mapping, which I have an issue on my plant for next week to do, I will document these columns. Is that okay for the moment?

Reply:
@cmgosnell I will do this as part of the prep-work outlined in #2040 for handing off the DBF to XBRL mapping for the remaining targeted XBRL tables.